lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -660,6 +644,81 @@ bool mlir::triton::preProcessLoopAndGetSchedule(
   return true;
 }
 
+/// Find the minimum number of insert_slice ops between the extract
+/// and the insert. This is equivalent to counting async_commit_group ops
+static int minWaitNumberForExtract(ttg::ExtractSliceOp extractOp,
+                                   scf::ForOp forOp) {
+  auto countInsertsBetween = [](Operation *op1, Operation *op2) {
+    int count = 0;
+    for (auto op = op1; op != op2; op = op->getNextNode()) {
+      if (isa<ttg::InsertSliceAsyncOp>(op))
+        count++;
+      // Intentionally skip block ops' children. This will give us
+      // convervatively low number of insert ops.
+    }
+    return count;
+  };
+
+  int minWaitNumber = INT_MAX;
+
+  // DFS the def chain of the extract op to find the insert op. On each path
+  // we calculate the number of insert ops. Then we select the minimum number
+  // of insert ops among all the paths.
+  std::function<int(Value, Operation *, int)> minOverHistories =
+      [&](Value val, Operation *sinkOp, int thisHistorySum) -> int {
+    if (auto insertOp = val.getDefiningOp<ttg::InsertSliceAsyncOp>()) {
+      thisHistorySum += countInsertsBetween(insertOp->getNextNode(), sinkOp);
+      minWaitNumber = std::min(minWaitNumber, thisHistorySum);
+      return minWaitNumber;
+    }
+    if (auto arg = val.dyn_cast<BlockArgument>()) {
+      auto block = arg.getOwner();
+      auto currForOp = dyn_cast<scf::ForOp>(block->getParentOp());
+
+      // We cannot track further. Conservatively return 0, and assert in debug,
+      // so that we can easier find unsupported cases.
+      assert(currForOp == forOp && "Unexpected forOp in minOverHistories.");
```

Comment:
If `currForOp != forOp`, it just throwed a runtime error. The following if condition never executes.

Reply:
And what if I have a nested for loop that I want to pipeline the inner one?

Reply:
> If `currForOp != forOp`, it just throws out a runtime error. The following if condition never executes.
My idea was that in Release we would simply return 0, as it is correct. But it also means that we have a case in the code for which we are generating probably very suboptimal code, so I have added an assert, so in Debug we clearly see that there is something to improve in the code. Does it make sense?

Reply:
> And what if I have a nested for loop that I want to pipeline the inner one?

This should work for basic cases, and is actually tested by one of the lit tests. The case that will work is when the pointers are not being passed around from one iteration of the outer loop to the next. If they are, we would hit this case (currForOp != forOp), because we would try to track the def chain of the `extract` through the outer loop.
It is possible that I could do better than returning `0` here, maybe the current `minWaitNumber` is an upper bound and we can return it, but I would need to think about it

Reply:
> If they are, we would hit this case (currForOp != forOp), because we would try to track the def chain of the extract through the outer loop.

Yes, that's what I'm referring to. 

> Debug we clearly see that there is something to improve in the code. Does it make sense?

We build with assert in the develop mode. (I think many assert should be replaced with "report_fatal_error" though)
If assert is triggered, it will actual cause many cases fail.

Reply:
OK, I'll remove the asserts. I won't replace them with `report_fatal_error`, because they IR is functionally correct. What we could use here is some kind of perf warning, but we don't have that.

Reply:
Sounds good!